File: src/java/com/springbok/pattern/PatternERR_007V01.java
/* Copyright (C) 2016 Springbok LLC <http://springbok.io/> All rights reserved. */

package com.springbok.pattern;

import com.springbok.utility.PatternUtility;

import java.util.Map;

/**
 * Describes the ITU antenna pattern ERR_007V01.
 */
public class PatternERR_007V01 implements EarthPattern, ReceivePattern {

	/** Maximum antenna gain [dB] */
	private double GainMax;

	/** Diameter of an earth antenna [m] */
	private double Diameter;

	/* Intermediate gain calculation results */
	private double lambda;
	private double d_over_lambda;
	private double G1;
	private double phi_m;
	private double phi_r;
	private double phi_b;

	/**
	 * Constructs a PatternERR_007V01 given an antenna maximum gain, and diameter.
	 *
	 * @param GainMax
	 *            Maximum antenna gain [dB]
	 * @param Diameter
	 *            Diameter of an earth antenna [m]
	 */
	public PatternERR_007V01(double GainMax, double Diameter) {

		/* Validate input parameters */
		PatternUtility.validate_input("GainMax", GainMax);
		PatternUtility.validate_input("Diameter", Diameter);

		/* Assign properties */
		this.GainMax = GainMax;
		this.Diameter = Diameter;

		/* Compute derived properties */
		// TODO: Speed of light should be a referenced constant
		this.lambda = 299792458 / (12.1 * 1e9);
		this.d_over_lambda = this.Diameter / this.lambda;
		this.phi_r = 95.0 / this.d_over_lambda;
		this.G1 = 29.0 - 25.0 * Math.log10(this.phi_r);
		this.phi_m = 20.0 / this.d_over_lambda * Math.sqrt(this.GainMax - this.G1);
		this.phi_b = Math.pow(10, 34.0 / 25.0);
	}

	/**
	 * Gets gainMax Maximum antenna gain [dB].
	 * 
	 * @return GainMax Maximum antenna gain [dB]
	 */
	public double getGainMax() {
		return GainMax;
	}

	/**
	 * Gets diameter of an earth antenna [m].
	 * 
	 * @return Diameter of an earth antenna [m]
	 */
	public double getDiameter() {
		return Diameter;
	}

	/**
	 * Gets wavelength of an earth antenna [m].
	 * 
	 * @return Wavelength of an earth antenna [m]
	 */
	public double getLambda() {
		return lambda;
	}

	/**
	 * Gets ratio of antenna diameter to wavelength.
	 * 
	 * @return Ratio of antenna diameter to wavelength
	 */
	public double getD_over_lambda() {
		return d_over_lambda;
	}

	/**
	 * Gets gain of the first side lobe (plateau constant part) [dB].
	 * 
	 * @return Gain of the first side lobe (plateau constant part) [dB]
	 */
	public double getG1() {
		return G1;
	}

	/**
	 * Gets start angle of the first side lobe (plateau constant part) [degrees].
	 * 
	 * @return Start angle of the first side lobe (plateau constant part) [degrees]
	 */
	public double getPhi_m() {
		return phi_m;
	}

	/**
	 * Gets start angle of the near side lobe (logarithmic part) [degrees].
	 * 
	 * @return Start angle of the near side lobe (logarithmic part) [degrees].
	 */
	public double getPhi_r() {
		return phi_r;
	}

	/**
	 * Gets start angle of the back lobe (back constant part) [degrees].
	 * 
	 * @return Start angle of the back lobe (back constant part) [degrees]
	 */
	public double getPhi_b() {
		return phi_b;
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see com.springbok.pattern.EarthPattern#copy()
	 */
	public PatternERR_007V01 copy() {
		return new PatternERR_007V01(this.GainMax, this.Diameter);
	}

	/**
	 * Appendix 30 (RR-2003) reference receiving earth station antenna pattern for
	 * Regions 1 and 3 (WRC-97).
	 *
	 * @param Phi
	 *            Angle for which a gain is calculated [degrees]
	 * @param options
	 *            Optional parameters (entered as key/value pairs)
	 * 
	 * @return Co-polar and cross-polar gain [dB]
	 */
	public Gain gain(double Phi, Map options) {

		/* Validate input parameters */
		PatternUtility.validate_input("Phi", Phi);

		/* Validate input options */
		options = PatternUtility.validateOptions(options);

		/* Implement pattern */
		double eps = Math.ulp(1.0);
		double phi = Math.max(eps, Phi);

		double G0 = GainMax - 2.5e-3 * Math.pow(d_over_lambda * phi, 2);
		double G2 = 29.0 - 25.0 * Math.log10(phi);
		double G3 = -5.0;
		double G4 = 0.0;

		double G = G0 * (0 <= phi && phi < phi_m ? 1 : 0) + G1 * (phi_m <= phi && phi < phi_r ? 1 : 0)
				+ G2 * (phi_r <= phi && phi < phi_b ? 1 : 0) + G3 * (phi_b <= phi && phi < 70 ? 1 : 0)
				+ G4 * (70 <= phi && phi <= 180 ? 1 : 0);

		double Gx0 = GainMax - 25.0;

		double phi_0 = 2.0 / d_over_lambda * Math.sqrt(3 / 0.0025);
		double Gx1 = GainMax - 25.0 + 8.0 * (phi - 0.25 * phi_0) / (0.19 * phi_0);
		double Gx2 = GainMax - 17.0;

		double phi_1 = phi_0 / 2.0 * Math.sqrt(10.1875);
		double S = 21.0 - 25.0 * Math.log10(phi_1) - (GainMax - 17.0);
		double Gx3 = GainMax - 17.0 + S * Math.abs((phi - phi_0) / (phi_1 - phi_0));
		double Gx4 = 21.0 - 25.0 * Math.log10(phi);
		double Gx5 = -5.0;
		double Gx6 = 0.0;

		double phi_2 = Math.pow(10.0, 26.0 / 25.0);

		double Gx = Gx0 * (0.0 <= phi && phi < 0.25 * phi_0 ? 1 : 0)
				+ Gx1 * (0.25 * phi_0 <= phi && phi < 0.44 * phi_0 ? 1 : 0)
				+ Gx2 * (0.44 * phi_0 <= phi && phi < phi_0 ? 1 : 0) + Gx3 * (phi_0 <= phi && phi < phi_1 ? 1 : 0)
				+ Gx4 * (phi_1 <= phi && phi < phi_2 ? 1 : 0) + Gx5 * (phi_2 <= phi && phi < 70.0 ? 1 : 0)
				+ Gx6 * (70.0 <= phi && phi <= 180.0 ? 1 : 0);

		/* Validate low level rules */
		if (0 < S) {
			throw new IllegalStateException("0 is less than S [6031: STDC_ERR_0_LT_S]");
		}
		if (phi_r < phi_m) {
			throw new IllegalStateException("phi_r is less than phi_m [6009: STDC_ERR_PHIR_LT_PHIM]");
		}
		if (phi_2 < phi_1) {
			throw new IllegalStateException("Phi_2 is less than Phi_1 [6008: STDC_ERR_ANG2_LT_ANG1]");
		}
		if (GainMax < G1) {
			throw new IllegalStateException("GainMax is less than G1 [6001: STDC_ERR_GNAX_LT_G1]");
		}
		if (phi_r < phi_m) {
			PatternUtility.logger.warn("phi_r is less than phi_m.");
		}

		/* Validate output parameters */
		if ((boolean) options.get("DoValidate")) {
			PatternUtility.validate_output(G, Gx, GainMax);
		}
		return new Gain(G, Gx);
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see java.lang.Object#hashCode()
	 */
	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		long temp;
		temp = Double.doubleToLongBits(GainMax);
		result = prime * result + (int) (temp ^ (temp >>> 32));
		temp = Double.doubleToLongBits(Diameter);
		result = prime * result + (int) (temp ^ (temp >>> 32));
		return result;
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see java.lang.Object#equals(java.lang.Object)
	 */
	@Override
	public boolean equals(Object obj) {
		if (this == obj) {
			return true;
		}
		if (obj == null) {
			return false;
		}
		if (!(obj instanceof PatternERR_007V01)) {
			return false;
		}
		PatternERR_007V01 other = (PatternERR_007V01) obj;
		if (Double.doubleToLongBits(GainMax) != Double.doubleToLongBits(other.GainMax)) {
			return false;
		}
		if (Double.doubleToLongBits(Diameter) != Double.doubleToLongBits(other.Diameter)) {
			return false;
		}
		return true;
	}
}

File: src/matlab/com/springbok/pattern/PatternERR_007V01.m
% Copyright (C) 2022 Springbok LLC
% 
% This program is free software: you can redistribute it and/or modify
% it under the terms of the GNU General Public License as published by
% the Free Software Foundation, either version 3 of the License, or (at
% your option) any later version.
% 
% This program is distributed in the hope that it will be useful, but
% WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
% General Public License for more details.
% 
% You should have received a copy of the GNU General Public License
% along with this program.  If not, see <https://www.gnu.org/licenses/>.
% 
classdef PatternERR_007V01 < EarthPattern & ReceivePattern
% Describes the ITU antenna pattern ERR_007V01.
  
  properties (SetAccess = private, GetAccess = public)
    
    % Maximum antenna gain [dB]
    GainMax
    % Diameter of an earth antenna, m
    Diameter
    
    % Intermediate gain calculation results
    lambda
    d_over_lambda
    G1
    phi_m
    phi_r
    phi_b
    
  end % properties (SetAccess = private, GetAccess = public)
  
  methods
    
    function this = PatternERR_007V01(GainMax, Diameter)
    % Constructs a PatternERR_007V01 given an antenna maximum gain, and diameter.
    %
    % Parameters
    %   GainMax - Maximum antenna gain [dB]
    %   Diameter - Diameter of an earth antenna, m
      
    % No argument constructor
      if nargin == 0
        return;
        
      end % if
      
      % Check number and class of input arguments.
      if ~(nargin == 2 ...
           && isnumeric(GainMax) ...
           && isnumeric(Diameter))
        exc = SException( ...
            'Springbok:InvalidInput', ...
            'Numeric GainMax, and Diameter must be provided.');
        throw(exc)
        
      end % if
      
      % Validate input parameters.
      PatternUtility.validate_input('GainMax', GainMax);
      PatternUtility.validate_input('Diameter', Diameter);
      
      % Assign properties
      this.GainMax = GainMax;
      this.Diameter = Diameter;
      
      % Compute derived properties
      this.lambda = 299792458 / (12.1 * 1e9);
      this.d_over_lambda = this.Diameter / this.lambda;
      this.phi_r = 95 / this.d_over_lambda;
      this.G1 = 29 - 25 * log10(this.phi_r);
      this.phi_m = 20 / this.d_over_lambda * sqrt(this.GainMax - this.G1);
      this.phi_b = 10 ^ (34 / 25);
      
    end % PatternERR_007V01()
    
    function that = copy(this)
    % Copies a PatternERR_007V01 given a maximum antenna gain.
    %
    % Parameters
    %   None
      that = PatternERR_007V01(this.GainMax, this.Diameter);

    end % copy()

    function [G, Gx, fH] = gain(this, Phi, varargin)
    % Appendix 30 (RR-2003) reference receiving earth station
    % antenna pattern for Regions 1 and 3 (WRC-97).
    %
    % Parameters:
    %   Phi - Angle for which a gain is calculated [degrees]
    %
    % Optional parameters (entered as key/value pairs):
    %   PlotFlag - Flag to indicate plot, true or {false}
    %
    % Returns:
    %   G - Co-polar gain [dB]
    %   Gx - Cross-polar gain [dB]
    %   fH - figure handle
      
    % Check number and class of input arguments.
      if ~(nargin >= 1 ...
           && isnumeric(Phi))
        exc = SException( ...
            'Springbok:InvalidInput', ...
            'Numeric Phi must be provided.');
        throw(exc)
        
      end % if
      
      % Validate input parameters.
      PatternUtility.validate_input('Phi', Phi);
      
      % Handle optional inputs.
      input = struct();
      if ~isempty(varargin)
        nVarargin = size(varargin, 2);
        
        if rem(nVarargin, 2)
          exc = SException( ...
              'Springbok:InvalidInput', ...
              'Inputs must be key/value pairs.');
          throw(exc)
          
        else
          for iVarargin = 1:2:nVarargin
            key = varargin{iVarargin};
            value = varargin{iVarargin + 1};
            input = PatternUtility.validate_input(key, value, input);
            
          end % for
          
        end % if
        
      end % if
      
      % Assign default values to any optional input parameters that
      % were not included in varargin.
      if ~isfield(input, 'PlotFlag')
        PlotFlag = false;

      else
        PlotFlag = input.PlotFlag;

      end % if
      if ~isfield(input, 'DoValidate')
        DoValidate = true;

      else
        DoValidate = input.DoValidate;

      end % if

      % Implement pattern.
      phi = max(eps, Phi); 
      
      G0 = this.GainMax - 2.5e-3 * (this.d_over_lambda .* phi) .^ 2;
      
      G2 = 29 - 25 * log10(phi);
      G3 = -5;
      G4 = 0;
      
      G = G0 .* (    0 <= phi & phi <  this.phi_m) ...
          + this.G1 .* (this.phi_m <= phi & phi <  this.phi_r) ...
          + G2 .* (this.phi_r <= phi & phi <  this.phi_b) ...
          + G3 .* (this.phi_b <= phi & phi <     70) ...
          + G4 .* (   70 <= phi & phi <=   180);
      
      Gx0 = this.GainMax - 25;
      
      phi_0 = 2 / this.d_over_lambda * sqrt(3 / 0.0025);
      Gx1 = this.GainMax - 25 + 8 * (phi - 0.25 * phi_0) / (0.19 * phi_0);
      Gx2 = this.GainMax - 17;
      
      phi_1 = phi_0 / 2 * sqrt(10.1875);
      S = 21 - 25 * log10(phi_1) - (this.GainMax - 17);
      Gx3 = this.GainMax - 17 + S * abs((phi - phi_0) / (phi_1 - phi_0));
      Gx4 = 21 - 25 * log10(phi);
      Gx5 = -5;
      Gx6 = 0;
      
      phi_2 = 10 ^ (26/25);
      
      Gx = Gx0 .* (0            <= phi & phi <  0.25 * phi_0) ...
           + Gx1 .* (0.25 * phi_0 <= phi & phi <  0.44 * phi_0) ...
           + Gx2 .* (0.44 * phi_0 <= phi & phi <         phi_0) ...
           + Gx3 .* (phi_0        <= phi & phi <         phi_1) ...
           + Gx4 .* (phi_1        <= phi & phi <         phi_2) ...
           + Gx5 .* (phi_2        <= phi & phi <            70) ...
           + Gx6 .* (70           <= phi & phi <=          180);
      
      % Validate low level rules.
      if 0 < S
        exc = SException( ...
            'Springbok:InvalidResult', ...
            '0 is less than S.', ...
            -6031, ...
            'STDC_ERR_0_LT_S', ...
            '0 is less than S.');
        throw(exc)
        
      end % if
      
      if this.phi_r < this.phi_m
        exc = SException( ...
            'Springbok:InvalidResult', ...
            'phi_r is less than phi_m.', ...
            -6009, ...
            'STDC_ERR_PHIR_LT_PHIM', ...
            'phi_r is less than phi_m.');
        throw(exc)
        
      end % if
      
      if phi_2 < phi_1
        exc = SException( ...
            'Springbok:InvalidResult', ...
            'Phi_2 is less than Phi_1.', ...
            -6008, ...
            'STDC_ERR_ANG2_LT_ANG1', ...
            'Phi_2 is less than Phi_1.');
        throw(exc)
        
      end % if
      
      if this.GainMax < this.G1
        exc = SException( ...
            'Springbok:InvalidResult', ...
            'GainMax is less than G1.', ...
            -6001, ...
            'STDC_ERR_GNAX_LT_G1', ...
            'GainMax is less than G1.');
        throw(exc)
        
      end % if
      
      if this.phi_r < this.phi_m
        warning( ...
            'Springbok:InvalidResult', ...
            'phi_r is less than phi_m.');
      end % if
      
      % Validate output parameters.
      if DoValidate
        PatternUtility.validate_output(G, Gx, this.GainMax);
      
      end % if
      
      % Plot results, if requested.
      if PlotFlag == true
        fH = figure;
        semilogx(phi, G, 'k')
        grid on
        hold on
        semilogx(phi, Gx, 'k')
        v = axis;
        axis([0.1 180 v(3) v(4)])
        title([strrep(mfilename, '_', '\_'), ' Reference Antenna Pattern'])
        xlabel('Phi (degrees)')
        ylabel('Gain (dB)')
        
      else
        fH = [];
        
      end % if
      
    end % gain()
    
  end % methods
  
end % classdef
